substrate/client/transaction-pool/src/fork_aware_txpool/dropped_watcher.rs

```diff
@@ -125,41 +242,64 @@ where
 		&mut self,
 		block_hash: BlockHash<C>,
 		event: ViewStreamEvent<C>,
-	) -> Option<ExtrinsicHash<C>> {
+	) -> Option<DroppedTransaction<ExtrinsicHash<C>>> {
 		trace!(
 			target: LOG_TARGET,
-			"dropped_watcher: handle_event: event:{:?} views:{:?}, ",
-			event,
+			"dropped_watcher: handle_event: event:{event:?} from:{block_hash:?} future_views:{:?} ready_views:{:?} stream_map views:{:?}, ",
+			self.future_transaction_views.get(&event.0),
+			self.ready_transaction_views.get(&event.0),
 			self.stream_map.keys().collect::<Vec<_>>(),
 		);
 		let (tx_hash, status) = event;
 		match status {
-			TransactionStatus::Ready | TransactionStatus::Future => {
-				self.transaction_states.entry(tx_hash).or_default().insert(block_hash);
+			TransactionStatus::Future => {
+				self.future_transaction_views.entry(tx_hash).or_default().insert(block_hash);
+			},
+			TransactionStatus::Ready | TransactionStatus::InBlock(..) => {
+				if let Some(mut views) = self.future_transaction_views.remove(&tx_hash) {
```

Comment:
DQ: Why do we move the entire list of views from future to ready here? This is just one view reporting the transaction as ready now right? I would expect that we only remove that one block hash from the future transactions view list.

Reply:
Good question.

It is about if we want to drop transaction when the last view referencing a transaction is dropped. 

And now for transaction that is future, and never was seen as ready on any fork, we want to remove it from the pool when the last view referencing transaction is gone. We basically don't want to keep future transaction too long. If the transaction is no longer in any view it was likely pushed out by some newly incoming transactions.

If the transaction was seen as ready on any fork, it means that it can be soon included into some new fork that can pop out. This is why in a `DroppedWatcher` I decided to treat such transaction as ready. Due to this we will keep this transaction in a mempool for a while - we will not send a dropped event to the pool.

I treat an internal mempool as a secondary, best-effort, buffer for transactions - meaning that it can hold ready transactions that are currently not referenced by any view (note that it applies to  the situation when limits are hit). When the blocks are built, those txs will be resubmitted and will get to the view. The other approach that could be simple would be to simply drop everything what is not referenced by views. Maybe it is a way to go. I'll give it more consideration when all groundwork is done.

Please also note that hitting limits on transaction pool is not a typical case. I would not expect hitting limits even during spamming tests. Hitting limits may occur during some attack or maybe when block production is stopped. This should not even happen when finality is stalled (becasue FinalityTimeout events shall be triggered and listeners will be dropped - still to be [tested](https://github.com/paritytech/polkadot-sdk/issues/5482)).

I am still here experimenting a bit, so it may happen that these code will get re-iterated.

Reply:
Okay I could follow this explanation, but its kind of complicated behaviour, maybe a comment in the code would help for unsuspecting developers :D